SphericalManifold -> PolarManifold

luca-heltai opened the pull request:
Still have to check all tests, but this is ready to review. Will add many changed output once this has been accepted as the "right" version of SphericalManifold.

bangerth commented on include/deal.II/grid/manifold.h:
```diff
@@ -273,13 +273,36 @@ class Manifold : public Subscriptor
   /**
+   * Return an intermediate point between two given
+   * points. Overloading this function allows the default recursive
+   * implementation of the method get_new_point() that takes a
```
I think the term "recursive" is not quite correct here since there is no function that calls itself. Maybe use the term "pairwise reduction" we use in the paper?

bangerth commented on include/deal.II/grid/manifold.h:
```diff
@@ -273,13 +273,36 @@ class Manifold : public Subscriptor
-   * In its default implementation it calls internally the function
-   * project_to_manifold. User classes can get away by simply implementing
+   * In its default implementation it calls recursively the function
+   * get_new_point(). User classes can get away by simply implementing
```
same here. maybe also say that it calls the `get_new_point()` variant with three arguments?

bangerth commented on include/deal.II/grid/manifold.h:
```diff
@@ -273,13 +273,36 @@ class Manifold : public Subscriptor
+   */
+  virtual
+  Point<spacedim>
+  get_new_point(  const Point<spacedim> &p1,
```
odd spaces around the parenthesis

bangerth commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -25,37 +25,37 @@
+ * Cartesian coordinates. In order for this manifold to work
+ * correctly, it cannot be attached to cells containing the center of
+ * the coordinate system or the north and south poles in three
+ * dimensions. These points are singular points of the coordinate
```
won't we have the same problem with the poles in 2d?

luca-heltai replied:
Actually no. In 2d we have periodicity, but no singularity. In other words, it is possible to write a single transformation with Jacobian different from zero everywhere (in 2D) (provided we allow for periodicity).

bangerth replied:
Oh yes, of course.

bangerth commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -127,6 +115,110 @@ class SphericalManifold : public ChartManifold<dim, spacedim, spacedim>
+ * While PolarManifold reflects the usual notion of polar coordinates,
+ * it may not be suitable for domains that contain either the north or
+ * south poles.
+ * Consider for istance the pair of points \f$x_1=(1,\pi/3,0)\f$ and
```
-> instance

bangerth commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -127,6 +115,110 @@ class SphericalManifold : public ChartManifold<dim, spacedim, spacedim>
+ * it may not be suitable for domains that contain either the north or
+ * south poles.
+ * Consider for istance the pair of points \f$x_1=(1,\pi/3,0)\f$ and
+ * \f$x_2=(1,\pi/3,\pi)\f$.
```
In which coordinate system do you consider these points? I suppose in polar coordinates, and in that case they would correspond to north and south pole, respectively, right? Maybe state so explicitly.

bangerth commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -127,6 +115,110 @@ class SphericalManifold : public ChartManifold<dim, spacedim, spacedim>
+ * connecting points on the surface of the sphere.
+ *
+ * This class implements a Manifold that joins any two points in space
+ * by first projecting them on the surface of a sphere with unit
```
on -> onto

bangerth commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -127,6 +115,110 @@ class SphericalManifold : public ChartManifold<dim, spacedim, spacedim>
+ * map of a sphere using only geodesic curves, and therefore, using
+ * this class with MappingManifold is discouraged. If you use this
+ * Manifold to describe the geometry of a sphere, you should use
+ * MappingQ as the underlying mapping, and not MappingManifold.
```
Oh, that's a shame. I was really looking forward to it. What is the reason?

luca-heltai replied:
I think @ESeNonFossiIo  committed a version of the paper in which this is explained in detail. The main reason is that if you take the average of three points, by pairwise simplification, this process dependent on the order in which you do it, i.e., average of p1, p2, p3 (with weight 1/3) and of p1, p3, p2 (with weight 1/3) do not commute if you construct the lines as geodesics.

bangerth replied:
By the way, I think this would read better if you stressed the point that the issue is using this class for a *whole* sphere. I think it should be totally fine to use with `MappingManifold` for a segment of the sphere.

Also, partly just to satisfy my curiosity, if one starts with a sufficiently fine mesh, the mapping only ever gets to see individual cell's share of the sphere. These are all locally smooth, with no singularities as long as you stay away from the center. Is this not enough to make things work?

bangerth commented on source/grid/manifold.cc:
```diff
@@ -44,21 +42,42 @@ project_to_manifold (const std::vector<Point<spacedim> > &,
-  Point<spacedim> p;
-  for (unsigned int i=0; i<quad.size(); ++i)
-    p += quad.weight(i) * quad.point(i);
+  QSorted<spacedim> sorted_quad(quad);
```
Ah, no way, we already have that class? What a fortuitous coincidence!

luca-heltai replied:
Coincidences do not exist...  :)

bangerth commented on source/grid/manifold_lib.cc:
```diff
@@ -186,6 +160,88 @@ SphericalManifold<dim,spacedim>::push_forward_gradient(const Point<spacedim> &sp
+  // Find the angle gamma described by v1 and v2:
+  const double gamma = std::acos((v1*v2)/(r1*r2));
+
+  // Find the angle sigma that correspont to archlengh equal to w
```
-> arclength

bangerth replied:
-> corresponds

bangerth commented on source/grid/manifold_lib.cc:
```diff
@@ -186,6 +160,88 @@ SphericalManifold<dim,spacedim>::push_forward_gradient(const Point<spacedim> &sp
+  // Find the angle sigma that correspont to archlengh equal to w
+  const double sigma = (1-w) * gamma;
+
+  // Versor with the same direction of v1
```
vector?

ESeNonFossiIo replied:
https://en.wikipedia.org/wiki/Versor_(physics)

bangerth replied:
Ha! I never heard of this word. Maybe replace it by "a unit vector in the direction of v1"?

bangerth commented on source/grid/manifold_lib.cc:
```diff
@@ -186,6 +160,88 @@ SphericalManifold<dim,spacedim>::push_forward_gradient(const Point<spacedim> &sp
+  const double r1 = (p1 - center).norm();
+  const double r2 = (p2 - center).norm();
+  const Tensor<1,spacedim> e1 = (p1 - center)/r1;
+  const Tensor<1,spacedim> e2 = (p2 - center)/r2;
```
what do you do if one of the points is the origin? Or close to the origin?

bangerth commented on source/grid/manifold_lib.cc:
```diff
@@ -186,6 +160,88 @@ SphericalManifold<dim,spacedim>::push_forward_gradient(const Point<spacedim> &sp
+
+  // Tangent vector to the unit sphere along the geodesic given by e1 and e2.
+  Tensor<1,spacedim> tg = (e2-(e2*e1)*e1);
+  tg = tg / tg.norm();
```
and what do you do if the two points lie radially from each other? In that case, the tangent vector needs to be radial as well.

ESeNonFossiIo replied:
Good point. In these case it does not work. I wrote these functions thinking to spacedim=3 dim=2. Luca ask me to generalise it to the case spacedim=3 and dim=3 but probably it is not easy as i thought.

ESeNonFossiIo replied:
I would like to use geodesic for <2,3> but these are not geodesic for <3,3>. May be my idea works only in the case we consider the sphere without the origin.

bangerth commented:
I haven't tried to look through the tests or testsuite changes, but like the overall idea. I'm unclear, however, how the new class deals with points at the origin (see my comments).

luca-heltai commented:
Currently failing tests:

~~~
804:mpi/fe_field_function_02.mpirun=3.debug
805:mpi/fe_field_function_02.mpirun=3.release
1663:manifold/spherical_manifold_02.debug
1664:manifold/spherical_manifold_02.release
1851:codim_one/mapping_fe_field_01.debug
1852:codim_one/mapping_fe_field_01.release
3256:fe/get_fe_from_name_01.debug
3257:fe/get_fe_from_name_01.release
6081:bits/get_fe_from_name.debug
6082:bits/get_fe_from_name.release
7108:slepc/step-36_parallel.mpirun=3.debug
7109:slepc/step-36_parallel.mpirun=3.release
7160:mappings/mapping_manifold_06.debug
7161:mappings/mapping_manifold_06.release
8137:grid/grid_generator_hyper_sphere.debug
8138:grid/grid_generator_hyper_sphere.release
8213:grid/torus_01.debug
8214:grid/torus_01.release
~~~

Some of these are failing because of a floating point exception in the new `SphericalManifold`, and some are failing due to a bug @ESeNonFossiIo  and I are hunting for in `MappingManifold`. Nearly there...

luca-heltai commented:
@bangerth : `SphericalManifold`, like `PolarManifold`, is still singular at the origin. It computes geodesics on codimension one surfaces, and then connects them linearly along the radii. Therefore it is *not* a geodesic in dim == spacedim. In fact, for dim=2, it is identical to `PolarManifold`, with the same problems. 

I'm open to suggestions on how to treat the dim==spacedim case in a different way on the origin. Maybe we should consider renaming this to `HyperShellManifold`, so that the name in itself contains somehow the limitation of the class...

luca-heltai commented:
Failing tests:

- [ ] mpi/fe_field_function_02.mpirun=3
- [ ] manifold/spherical_manifold_02
- [ ] codim_one/mapping_fe_field_01
- [ ] fe/get_fe_from_name_01
- [ ] bits/get_fe_from_name
- [ ] slepc/step-36_parallel.mpirun=3
- [ ] mappings/mapping_manifold_06
- [ ] grid/grid_generator_hyper_sphere
- [ ] grid/torus_01

ESeNonFossiIo commented:
I can not modify your post. 
Following tests are ok:
- *codim_one/mapping_fe_field_01*
- *grid/grid_generator_hyper_sphere* 
- *fe/get_fe_by_name_01*

luca-heltai commented:
This is now ready to be reviewed. There was a subtle bug in the recursive implementation of `get_new_point`. We only have one failing test, which seems unrelated to this PR, as it started failing after the rebase:

~~~

7106: --------------------------------------------------------
7106: An error occurred in line <1160> of file <../source/dofs/dof_tools.cc> in function
7106:     std::vector<IndexSet> dealii::DoFTools::locally_owned_dofs_per_subdomain(const DoFHandlerType &) [DoFHandlerType = dealii::DoFHandler<2, 2>]
7106: The violated condition was: 
7106:     n_subdomains > *std::max_element (subdomain_association.begin (), subdomain_association.end ())
7106: The name and call sequence of the exception was:
7106:     ExcInternalError()
7106: Additional Information: 
7106: This exception -- which is used in many places in the library -- usually indicates that some condition which the author of the code thought must be satisfied at a certain point in an algorithm, is not fulfilled. An example would be that the first part of an algorithm sorts elements of an array in ascending order, and a second part of the algorithm later encounters an an element that is not larger than the previous one.
7106: 
7106: There is usually not very much you can do if you encounter such an exception since it indicates an error in deal.II, not in your own program. Try to come up with the smallest possible program that still demonstrates the error and contact the deal.II mailing lists with it to obtain help.
7106: --------------------------------------------------------
~~~

tamiko commented:
/run-tests

tamiko commented:
@luca-heltai What about I run the full testsuite against this monster +100k -100k changeset?

tamiko commented:
Otherwise :+1:

luca-heltai commented:
I did run the full testsuite, and it passes all except one test, unrelated to this patch...

tamiko commented:
@luca-heltai I thought about running the full simserv04 testsuite :-]

@bangerth I suggest we merge.

tamiko commented:
@luca-heltai Please rebase and merge.

bangerth commented on include/deal.II/grid/manifold.h:
```diff
@@ -125,15 +125,19 @@ namespace Manifolds
+ * only overload Manifold::project_to_manifold() for simple
+ * situations. This is often useful when describing manifolds that are
+ * embedded in higher dimensional space, e.g., the surface of a
+ * sphere.  In those cases, the desired new point maybe computed
```
if you're still accepting comments: maybe -> may be

bangerth commented on include/deal.II/grid/manifold.h:
```diff
@@ -125,15 +125,19 @@ namespace Manifolds
+ * situations. This is often useful when describing manifolds that are
+ * embedded in higher dimensional space, e.g., the surface of a
+ * sphere.  In those cases, the desired new point maybe computed
+ * simply by the (weighted) average of the provided point, projected
```
point -> points

bangerth commented on include/deal.II/grid/manifold.h:
```diff
@@ -273,18 +277,48 @@ class Manifold : public Subscriptor
+   * In its default implementation it uses a pair-wise reduction of
+   * the points in the quadrature formula by calling the function
+   * get_intermediate_point() on the first two points, then on the
+   * resulting point and the next, untill all points in the quadrature
```
-> until

bangerth commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -127,6 +116,131 @@ class SphericalManifold : public ChartManifold<dim, spacedim, spacedim>
+ *
+ * You can use this Manifold object to describe any sphere, circle,
+ * hypersphere or hyperdisc in two or three dimensions, both as a co-dimension
+ * one manifold descriptor or as co-dimension zero manifold descriptor.
```
are you saying that this can be used for "a co-dimension one manifold descriptor of a spherical surface embedded in a higher dimensional space, or as a co-dimension zero manifold descriptor for a body with positive volume"? I *think* that's what you're trying to say, but it's not entirely clear to me.

bangerth commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -127,6 +116,131 @@ class SphericalManifold : public ChartManifold<dim, spacedim, spacedim>
+ * pole, the south pole or the center.  Consider for instance the pair
+ * of points \f$x_1=(1,\pi/3,0)\f$ and \f$x_2=(1,\pi/3,\pi)\f$ in
+ * polar coordinates (laying on the surface of a sphere with radius
+ * one, on a parallel at height $\pi/3$. In this case connecting the
```
missing closing parenthesis

bangerth replied:
btw, it should be "lying" not "laying". The former is passive (the point is there) whereas the latter is active (the point is being put there, as in "laying bricks")

bangerth commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -127,6 +116,131 @@ class SphericalManifold : public ChartManifold<dim, spacedim, spacedim>
+ * where $\kappa = \frac{x_1 \times \x_2}{\Vert x_1 \times \x_2 \Vert}$
+ * and $\alpha(t) = t * \arccos(x_1 * x_2) $ for $t\in[0,1]$.
+ * Indeed, this is a geodesic, and it is the natural choice when
+ * connecting points on the surface of the sphere.
```
I *think* what you're trying to say with this example is that the latter is what the current class actually implements. Maybe say somewhere in this example that the former is what `PolarManifold` implements, and the latter what `SphericalManifold` implements?

bangerth commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -127,6 +116,131 @@ class SphericalManifold : public ChartManifold<dim, spacedim, spacedim>
+ *
+ * If the codimension of the Manifold is one, than this Manifold
+ * connects points using geodesics. In all other cases it is a
+ * continuus extension of the codimension one case.
```
-> continuous

What does this actually mean? That it finds the geodesic in (phi,theta) space, and then continuously interpolates this geodesic in radial direction between the two points?

bangerth replied:
ah, forget the second part. This is explained in the next paragraph

bangerth commented:
I had only a couple more small comments on documentation. Please merge yourself after addressing as you see fit.

luca-heltai commented:
:) There's still a lot to do... but we're on the right track!

bangerth commented:
I thought about this a bit yesterday but couldn't quite wrap my mind around the issue with the center of the `SphericalManifold` class. Are the only problematic lines the ones that go straight through the center? Or are there other issues you see that would prevent the use for full spheres?

tamiko commented:
@luca-heltai We have a bit of a fallout on the tester [1].

[1] http://cdash.kyomu.43-1.org/viewTest.php?onlydelta&buildid=516

luca-heltai commented:
The only problematic lines are those through the center. If a cell contains the center of the manifold, then you should not set its `manifold_id` to `SphericalManifold`.

luca-heltai commented:
@tamiko, I'm looking at those right now. The full test suite runs fine on three of my machines... :(

bangerth commented:
@luca-heltai -- could we special case these lines? I.e., if `x1 * x2   \approx   - x1.norm()* x2.norm()`, then we know that the line goes straight through the origin, and we just take the weighted average of `x1` and `x2` in Cartesian coordinates.

bangerth commented:
It's a good question, though, whether that works if you have, say, 4 points at the vertices of a square centered at the origin, and you do the pairwise reduction. In fact, that's not going to work if you considered the points in clockwise orientation -- every point is at the same radius, so you get a point on that radius, rather than the origin, if you average them. Bummer :-(

luca-heltai commented:
yes, that's the worst example. There's simply no easy way to define a polar / spherical manifold in that case. You'd better off using a flat manifold for the central cell, and maybe just use polar / spherical manifold on the faces of that cell. Although in this case, if it is a single cell, you'd end up upon refinement with 4 corners of almost pi (not good). 

I don't know the answer to this. I think that for a full sphere, the best option remains to set the manifold only on the boundary. If you have hyper shells, instead, then the story is entirely different.

tamiko commented:
@luca-heltai I will verify the test results on simserv04 later today by running it on my laptop.

luca-heltai commented:
@tamiko, I don't know what happened, but now I see the same tests failing on one of my machines. I'm trying to figure out if I messed something up during the rebase, and committed things that I should not have.

bangerth commented:
@luca-heltai -- I agree that this may simply not work. Will have to think about this some more.

bangerth commented:
There was also at least one test that fell into an Assert (a normal vector was computed to be zero).

luca-heltai commented:
Yes. We added several asserts in `Manifold`, `ChartManifold`, and `PolarManifold`, resulting in a lot of assert in several tests that include singular points. These are all non-sensical tests, and should be removed/changed. I'm now checking that the ones that fail are actually of this type. PolarManifold cannot be used for full Spheres/Hyperspheres/HyperShells unless one removes the cells containing the north and south poles. 

@ESeNonFossiIo checked all of these, and fixed them, but I might have messed up during one of the rebases...

luca-heltai commented:
@tamiko , I found the offending rebase, and I'm now trying to restore things as @ESeNonFossiIo had them before my mess.

PR coming soon.

tamiko commented:
@luca-heltai 
>  most of the output files are simply not updated to reflect the changes in SphericalManifold.

That's at least something with an easy fix. :-)

kronbichler commented on source/grid/manifold.cc:
```diff
@@ -44,21 +42,42 @@ project_to_manifold (const std::vector<Point<spacedim> > &,
+      if ( w != 0 )
+        p = get_intermediate_point(p, sorted_quad.point(i) , w/(sorted_quad.weight(i) + w) );
+      w += sorted_quad.weight(i);
+    }
```
Since I'm just working on #2896 where I want to avoid those kind of changes where we allocate new classes: Could someone explain why we need to sort here? My idea for a change would be to do a linear search with quadratic complexity up to say 16 points (we should never have more than 8) and only afterwards go to the memory allocation. Any comments?

bangerth replied:
Yes, that would make perfect sense.

(One could also wonder what would happen if `Quadrature` *automatically* sorted points by weight, or triggered by a flag. Would this improve accuracy?)

luca-heltai replied:
Here sorting is required because it makes the `get_new_point` function single valued. In general we do not require associativity for the `get_intermediate_point` function, and sorting is a way to have a more stable numerical algorithm, that always returns the same value, independently on the order of the points.
